<div style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h1 style = "padding: 15px; 
                 color: White;
                 text-align:center;
                 font-family: Trebuchet MS;">Netflix Movies and TV Shows
    </h1>
</div>

<div align="center"> 
    <img src="https://content.api.news/v3/images/bin/3fcf3cdcc2cfd6905e582a186e46e584" alt="Netflix" width="75%" style="margin-top:2rem;margin-bottom:2rem">
</div>

<div>
    <p>Netflix is one of the most popular media and video streaming platforms. They have over 8000 movies or tv shows available on their platform, as of mid-2021, they have over 200M Subscribers globally.</p>
</div>

<div>
    <p>Analysis ideas:</p>
    <ul>
        <li><p>Understanding what content is available in different countries</p>
        <li><p>Identifying similar content by matching text-based features</p>
        <li><p>Network analysis of Actors / Directors and find interesting insights</p>
        <li><p>Does Netflix has more focus on TV Shows than movies in recent years.</p>
    </ul>
</div>

<div style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">Table of Contents
    </h2>
</div>

* [<span style="font-family: Trebuchet MS; font-size:15px;">1. Imports</span>](#imports)
* [<span style="font-family: Trebuchet MS; font-size:15px;">2. Loading the Dataset</span>](#loading-the-dataset)
* [<span style="font-family: Trebuchet MS; font-size:15px;">3. Understanding the Data</span>](#understanding-the-data)
    * [<span style="font-family: Trebuchet MS; font-size:15px;">3.1 Checking Null Values</span>](#checking-null-values)


<div id="imports"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">1. Imports
    </h2>
</div>

In [1]:
%%capture
import numpy as np
import pandas as pd
import plotly.express as px

<div id="loading-the-dataset"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">2. Loading the Dataset
    </h2>
</div>

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/sahinozan/Netflix-Movies-TV-Shows/main/netflix_titles.csv", encoding="utf-8")
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

<div id="understanding-the-data"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">3. Understanding the Data
    </h2>
</div>

Dataset consists of 12 features.
1. **show_id**: The unique id number for each show or movie
    - s1
    - s2
    - s3
    - ...
2. **type**: Type of the content
    - Movie
    - TV Show
3. **title**: Title of the content
    - The Starling
    - Squid Game
    - Jaws: The Revenge
    - ...
4. **director**: Name of the director for that movie or show
    - Steven Spielberg
    - Cédric Jimenez
    - Hirotsugu Kawasaki
    - ...
5. **cast**: Names of the cast for that movie or show 
    - Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane...
    - Antti Pääkkönen, Heljä Heikkinen, Lynne Guaglione, Pasi Ruohonen, Rauno Ahonen...
    - Luna Wedler, Jannis Niewöhner, Milan Peschel, Edin Hasanović...
    - ...
6. **country**: List of countries in which the content is available 
    - United States, South Africa, India...
    - United States, United Kingdom, Canada, Germany...
    - Canada, France, Japan, Russia...
    - ...
7. **data_added**: The date the content was added
    - September 25, 2021
    - September 22, 2021
    - August 23, 2021
    - ...
8. **release_year**: Release year of the content
    - 2020
    - 2014
    - 2006
    - ...
9. **rating**: Motion picture content rating 
    - PG-13
    - TV-14
    - R
    - ...
10. **duration**: Total duration of the content
    - 90 min
    - 2 seasons
    - 48 min
    - ...
11. **listed_in**: List of categories where content is listed
    - Documentaries, International TV Shows, TV Dramas, TV Mysteries...
    - Comedies, Dramas, International Movies...
    - Documentaries, TV Dramas, Comedies...
    - ...
12. **description**: Short description about the content
    - As her father nears the end of his life, filmmaker Kirsten Johnson stages his death in inventive and comical ways to help them both face the inevitable
    - After crossing paths at a party, a Cape Town teen sets out to prove whether a private-school swimming star is her sister who was abducted at birth
    - Tayo speeds into an adventure when his friends get kidnapped by evil magicians invading their city in search of a magical gemstone
    - ...

<div id="checking-null-values"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h3 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">3.1 Checking Null Values
    </h3>
</div>

Let's first check if there are any null values in the dataset

In [3]:
df.isna().sum().sort_values(ascending=False)

director        2634
country          831
cast             825
date_added        10
rating             4
duration           3
show_id            0
type               0
title              0
release_year       0
listed_in          0
description        0
dtype: int64

* There are small amount of null values in the **date_added**, **rating**, **duration** columns. 
* Meanwhile, there are large amount of null values in the **country**, **cast**, **director** columns.

<div id="optimizing-the-dataset"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h3 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">3.2 Optimizing the Dataset
    </h3>
</div>

Many of the features are `object` type. We can convert those features into more specific and compact types to save memory and increase efficiency.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


Let's first check the number of unique values for each feature to make the conversion process easier

In [5]:
for i in df.columns:
    print(f"Number of unique values in the {i} column: {df[i].nunique()}")

Number of unique values in the show_id column: 8807
Number of unique values in the type column: 2
Number of unique values in the title column: 8807
Number of unique values in the director column: 4528
Number of unique values in the cast column: 7692
Number of unique values in the country column: 748
Number of unique values in the date_added column: 1767
Number of unique values in the release_year column: 74
Number of unique values in the rating column: 17
Number of unique values in the duration column: 220
Number of unique values in the listed_in column: 514
Number of unique values in the description column: 8775


Some of the columns contain less unique values. These will be the ones that we do conversion on them. Columns that contain less unique values:
* **type**
* **release_year**
* **rating**

Other interesting features are **show_id** and **title**. They contain unique values for each row. I will check them first to see what is going.

In [6]:
print("Number of unique values:", df["show_id"].nunique())
print(df["show_id"].unique())

Number of unique values: 8807
['s1' 's2' 's3' ... 's8805' 's8806' 's8807']


There is a unique value for each show. I don't understand the point of this feature. It only contains the values starting from 1 to 8807 which is basically an index for every row. The problem is that we already have that information stored in our dataframe so, I will remove this feature because it doesn't contain any valuable information for the analysis.

In [7]:
df.drop(labels="show_id", axis=1, inplace=True)

In [8]:
print("Number of unique values:", df["title"].nunique())
print(df["title"].unique())

Number of unique values: 8807
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']


Similar to **show_id**, this column also contains unique values for each show but the difference is that these values are not just integers like an index value. These values contain the name of the each show which is probably will be used in the analysis. Therefore, I will not remove this feature.

In [9]:
print("Number of unique values:", df["type"].nunique())
print(df["type"].unique())

Number of unique values: 2
['Movie' 'TV Show']


There are only two unique values in this column. These are essentially used to categorize the content into **Movie** or **TV-Show**. I will convert this feature into `category`.

In [10]:
df["type"] = df["type"].astype("category")

In [11]:
print("Number of unique values:", df["release_year"].nunique())
print(df["release_year"].unique())

Number of unique values: 74
[2020 2021 1993 2018 1996 1998 1997 2010 2013 2017 1975 1978 1983 1987
 2012 2001 2014 2002 2003 2004 2011 2008 2009 2007 2005 2006 1994 2015
 2019 2016 1982 1989 1990 1991 1999 1986 1992 1984 1980 1961 2000 1995
 1985 1976 1959 1988 1981 1972 1964 1945 1954 1979 1958 1956 1963 1970
 1973 1925 1974 1960 1966 1971 1962 1969 1977 1967 1968 1965 1946 1942
 1955 1944 1947 1943]


There are 74 unique values in this column. Also, these are integer values. Smallest integer datatype that can hold these values is `int16`.

In [12]:
df["release_year"] = df["release_year"].astype("int16")

In [13]:
print("Number of unique values:", df["rating"].nunique())
print(df["rating"].unique())

Number of unique values: 17
['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']


There are 17 unique values in this column. These values contain both integers and strings. Therefore, we will convert this column into `category` type.

In [14]:
df["rating"] = df["rating"].astype("category")

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   type          8807 non-null   category
 1   title         8807 non-null   object  
 2   director      6173 non-null   object  
 3   cast          7982 non-null   object  
 4   country       7976 non-null   object  
 5   date_added    8797 non-null   object  
 6   release_year  8807 non-null   int16   
 7   rating        8803 non-null   category
 8   duration      8804 non-null   object  
 9   listed_in     8807 non-null   object  
 10  description   8807 non-null   object  
dtypes: category(2), int16(1), object(8)
memory usage: 585.8+ KB
